common/middleware/src/meta_rules.rs

```diff
@@ -0,0 +1,710 @@
+static LOGDNA_META_LABELS: &str = "LOGDNA_META_LABELS";
+
+static K8S_LOG_DIR: &str = "/var/log/containers/";
+
+static MAP_PREFIX_ENV: &str = "env.";
+static MAP_PREFIX_ANN: &str = "annot.";
+static MAP_PREFIX_LAB: &str = "label.";
+
+lazy_static! {
+    static ref REGEX_VAR: Regex = Regex::new(r"(?P<var>\$\{(?P<key>[^|}]+?)})").unwrap();
+    static ref REGEX_VAR_DEFAULT: Regex =
+        Regex::new(r"(?P<var>\$\{(?P<key>[^|}]+?)\|(?P<default>[^|}]*?)})").unwrap();
+    static ref REGEX_CRIO_LOG: RegexB =
+        RegexB::new(r"([0-9]{4}(?:-[0-9]{2}){2}T(?:[0-9]{2}:){2}[0-9]{2}.[0-9]{1,9}(?:[+-][0-9]{2}:[0-9]{2}|Z)) (stdout|stderr) ([PF]) (?P<line>.*)").unwrap();
+}
+
+//TODO: extract to LogConfig
+#[derive(Default)]
+pub struct MetaRulesConfig {
+    pub app: Option<String>,
+    pub host: Option<String>,
+    pub env: Option<String>,
+    pub file: Option<String>,
+    pub k8s_file: Option<String>, // for k8s lines, applied after "file"
+    pub meta: Option<String>,
+    pub annotations: Option<String>,
+    pub labels: Option<String>,
+}
+
+impl MetaRulesConfig {
+    pub fn from_env() -> Self {
+        let vars = os_env_hashmap();
+        MetaRulesConfig {
+            app: vars.get(LOGDNA_META_APP).cloned(),
+            host: vars.get(LOGDNA_META_HOST).cloned(),
+            env: vars.get(LOGDNA_META_ENV).cloned(),
+            file: vars.get(LOGDNA_META_FILE).cloned(),
+            k8s_file: vars.get(LOGDNA_META_K8S_FILE).cloned(),
+            meta: vars.get(LOGDNA_META_JSON).cloned(),
+            annotations: vars.get(LOGDNA_META_ANNOTATIONS).cloned(),
+            labels: vars.get(LOGDNA_META_LABELS).cloned(),
+        }
+    }
+}
+
+pub struct MetaRules {
+    env_map: HashMap<String, String>,
+    // "override" fields
+    over_app: Option<String>,
+    over_host: Option<String>,
+    over_env: Option<String>,
+    over_file: Option<String>,
+    over_k8s_file: Option<String>, // k8s lines only
+    over_meta: Option<String>,
+    over_annotations: Option<HashMap<String, String>>, // "merge override" (for "delete and then override" use disable k8s enrichment)
+    over_labels: Option<HashMap<String, String>>,      // --
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum MetaRulesError {
+    #[error(transparent)]
+    Serde(#[from] serde_json::Error),
+}
+
+impl MetaRules {
+    pub fn new(cfg: MetaRulesConfig) -> Result<MetaRules, MetaRulesError> {
+        let obj = MetaRules {
+            env_map: os_env_hashmap(),
+            over_app: cfg.app,
+            over_host: cfg.host,
+            over_env: cfg.env,
+            over_file: cfg.file,
+            over_k8s_file: cfg.k8s_file,
+            over_meta: cfg.meta,
+            over_annotations: cfg.annotations.map_or_else(
+                || None,
+                |str| match serde_json::from_str(str.as_str()) {
+                    Ok(kvp) => kvp,
+                    Err(err) => {
+                        panic!(
```

Comment:
If possible I'd prefer to propagate an error up to main. Users aren't used to seeing "panics" in logs and don't necessarily have the experience to parse them, which leads to more support calls.

If we pass it up to main as a normal error value then the existing code will already log it and exit the process

Reply:
+1. updated.